<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#一、收集数据" data-toc-modified-id="一、收集数据-1">一、收集数据</a></span></li><li><span><a href="#二、评估数据" data-toc-modified-id="二、评估数据-2">二、评估数据</a></span><ul class="toc-item"><li><span><a href="#（一）目视评估：寻找线索" data-toc-modified-id="（一）目视评估：寻找线索-2.1">（一）目视评估：寻找线索</a></span></li><li><span><a href="#（二）编程评估：深入探索" data-toc-modified-id="（二）编程评估：深入探索-2.2">（二）编程评估：深入探索</a></span><ul class="toc-item"><li><span><a href="#1.-twitters-表格" data-toc-modified-id="1.-twitters-表格-2.2.1">1. <strong><code>twitters</code> 表格</strong></a></span></li><li><span><a href="#2.-images-表格" data-toc-modified-id="2.-images-表格-2.2.2">2. <strong><code>images</code> 表格</strong></a></span></li><li><span><a href="#3.-twitter_additions-表格" data-toc-modified-id="3.-twitter_additions-表格-2.2.3">3. <strong><code>twitter_additions</code> 表格</strong></a></span></li></ul></li><li><span><a href="#（三）评估结果：问题记录" data-toc-modified-id="（三）评估结果：问题记录-2.3">（三）评估结果：问题记录</a></span><ul class="toc-item"><li><span><a href="#1.-数据质量" data-toc-modified-id="1.-数据质量-2.3.1">1. 数据质量</a></span></li><li><span><a href="#2.-数据结构" data-toc-modified-id="2.-数据结构-2.3.2">2. 数据结构</a></span></li></ul></li></ul></li><li><span><a href="#三、清理数据" data-toc-modified-id="三、清理数据-3">三、清理数据</a></span><ul class="toc-item"><li><span><a href="#（一）结构问题：补充表images、twitter_additions到表twitters中" data-toc-modified-id="（一）结构问题：补充表images、twitter_additions到表twitters中-3.1">（一）结构问题：补充表images、twitter_additions到表twitters中</a></span><ul class="toc-item"><li><span><a href="#1.-定义" data-toc-modified-id="1.-定义-3.1.1">1. 定义</a></span></li><li><span><a href="#2.-代码" data-toc-modified-id="2.-代码-3.1.2">2. 代码</a></span></li><li><span><a href="#3.-测试" data-toc-modified-id="3.-测试-3.1.3">3. 测试</a></span></li></ul></li><li><span><a href="#（二）结构问题：text列中的正文内容应和短网址分开，拆分为正文和网址两列" data-toc-modified-id="（二）结构问题：text列中的正文内容应和短网址分开，拆分为正文和网址两列-3.2">（二）结构问题：text列中的正文内容应和短网址分开，拆分为正文和网址两列</a></span><ul class="toc-item"><li><span><a href="#1.-定义" data-toc-modified-id="1.-定义-3.2.1">1. 定义</a></span></li><li><span><a href="#2.-代码" data-toc-modified-id="2.-代码-3.2.2">2. 代码</a></span></li><li><span><a href="#3.-测试" data-toc-modified-id="3.-测试-3.2.3">3. 测试</a></span></li></ul></li><li><span><a href="#（三）结构问题：doggo、floofer、pupper、puppo应合并为一列stage" data-toc-modified-id="（三）结构问题：doggo、floofer、pupper、puppo应合并为一列stage-3.3">（三）结构问题：doggo、floofer、pupper、puppo应合并为一列stage</a></span><ul class="toc-item"><li><span><a href="#1.定义" data-toc-modified-id="1.定义-3.3.1">1.定义</a></span></li><li><span><a href="#2.代码" data-toc-modified-id="2.代码-3.3.2">2.代码</a></span></li><li><span><a href="#3.测试" data-toc-modified-id="3.测试-3.3.3">3.测试</a></span></li></ul></li><li><span><a href="#（四）有效性问题：缺少p4以及相关的列" data-toc-modified-id="（四）有效性问题：缺少p4以及相关的列-3.4">（四）有效性问题：缺少p4以及相关的列</a></span><ul class="toc-item"><li><span><a href="#1.定义" data-toc-modified-id="1.定义-3.4.1">1.定义</a></span></li><li><span><a href="#2.代码" data-toc-modified-id="2.代码-3.4.2">2.代码</a></span></li><li><span><a href="#3.测试" data-toc-modified-id="3.测试-3.4.3">3.测试</a></span></li></ul></li><li><span><a href="#（五）有效性问题：列rating_numerator、rating_denominator出现过大或过小值" data-toc-modified-id="（五）有效性问题：列rating_numerator、rating_denominator出现过大或过小值-3.5">（五）有效性问题：列<em>rating_numerator、rating_denominator</em>出现过大或过小值</a></span><ul class="toc-item"><li><span><a href="#1.定义" data-toc-modified-id="1.定义-3.5.1">1.定义</a></span></li><li><span><a href="#2.代码" data-toc-modified-id="2.代码-3.5.2">2.代码</a></span></li><li><span><a href="#3.测试" data-toc-modified-id="3.测试-3.5.3">3.测试</a></span></li></ul></li><li><span><a href="#（六）有效性问题：列p1_dog为False，不属于狗" data-toc-modified-id="（六）有效性问题：列p1_dog为False，不属于狗-3.6">（六）有效性问题：列<em>p1_dog</em>为False，不属于狗</a></span><ul class="toc-item"><li><span><a href="#1.定义" data-toc-modified-id="1.定义-3.6.1">1.定义</a></span></li><li><span><a href="#2.代码" data-toc-modified-id="2.代码-3.6.2">2.代码</a></span></li><li><span><a href="#3.测试" data-toc-modified-id="3.测试-3.6.3">3.测试</a></span></li></ul></li><li><span><a href="#（七）准确性问题" data-toc-modified-id="（七）准确性问题-3.7">（七）准确性问题</a></span><ul class="toc-item"><li><span><a href="#1.定义" data-toc-modified-id="1.定义-3.7.1">1.定义</a></span></li><li><span><a href="#2.代码" data-toc-modified-id="2.代码-3.7.2">2.代码</a></span></li><li><span><a href="#3.测试" data-toc-modified-id="3.测试-3.7.3">3.测试</a></span></li></ul></li><li><span><a href="#（八）一致性问题" data-toc-modified-id="（八）一致性问题-3.8">（八）一致性问题</a></span><ul class="toc-item"><li><span><a href="#1.定义" data-toc-modified-id="1.定义-3.8.1">1.定义</a></span></li><li><span><a href="#2.代码" data-toc-modified-id="2.代码-3.8.2">2.代码</a></span></li><li><span><a href="#3.测试" data-toc-modified-id="3.测试-3.8.3">3.测试</a></span></li></ul></li><li><span><a href="#补充" data-toc-modified-id="补充-3.9">补充</a></span></li></ul></li></ul></div>

# 一、收集数据

In [2]:
import pandas as pd
import numpy as np

import requests
import os
import json

% matplotlib inline
import matplotlib.pyplot as plt

# 解决内容显示不全的问题
pd.set_option('max_colwidth', 200)

**1、导入给定的已下载WeRateDogs推特数据，形成DataFrame数据`twitters`**

In [30]:
twitters = pd.read_csv('twitter-archive-enhanced.csv')

2、下载给定网址的推特图像预测数据

In [17]:
url = 'https://raw.githubusercontent.com/udacity/new-dand-advanced-china/master/%E6%95%B0%E6%8D%AE%E6%B8%85%E6%B4%97/WeRateDogs%E9%A1%B9%E7%9B%AE/image-predictions.tsv'
response = requests.get(url)
with open(url.split('/')[-1], mode='wb') as file:
    file.write(response.content)    

**导入推荐图像预测数据，形成DataFrame数据`images`**

In [5]:
images = pd.read_csv('image-predictions.tsv', sep='\t')

3、导入项目提供的tweet_json.txt文件  
读入tweet ID、retweet_count 和 favorite_count 字段，形成DataFrame数据`twitter_additions`
并保存为`twitter_additions.csv`

In [3]:
twitter_additions_list = []
with open('tweet_json.txt', encoding='utf-8', mode='r') as file:
    for line in file:    
        parsed_json = json.loads(line)
        twitter_additions_list.append({'tweet_ID':parsed_json['id'],
                                       'retweet_count':parsed_json['retweet_count'],
                                       'favorite_count':parsed_json['favorite_count']})
twitter_additions = pd.DataFrame(twitter_additions_list, columns = ['tweet_ID', 'retweet_count', 'favorite_count'])
twitter_additions.to_csv('twitter_additions.csv')

**重新导入`twitter_additions.csv`**

In [7]:
twitter_additions = pd.read_csv('twitter_additions.csv')

# 二、评估数据

**关键要点**  

清洗这个项目的数据时要牢记几个要点：  
- 我们只需要含有图片的原始评级 (不包括转发)。尽管数据集中有 5000 多条数据，但是并不是所有都是狗狗评分，并且其中有一些是转发。
- 完整地评估和清理整个数据集将需要大量时间，实践和展示数据处理技巧没有必要将这个数据集全部清理。因此，本项目的要求只是评估和清理此数据集中的至少 8 个质量问题和至少 2 个整洁度问题。
- 根据 整洁数据 tidy data 的规则要求，本项目的数据清理应该包括将三个数据片段进行合并。
- 如果分子评级超过分母评级，不需要进行清洗。这个 特殊评分系统 是 WeRateDogs 人气度较高的主要原因。（同样，也不需要删除分子小于分母的数据）
- 不必收集 2017 年 8 月 1 日之后的数据，你可以收集到这些推特的基本信息，但是你不能收集到这些推特对应的图像预测数据，因为你没有图像预测算法的使用权限。

## （一）目视评估：寻找线索

In [8]:
twitters

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [9]:
images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [10]:
twitter_additions

Unnamed: 0            tweet_ID  retweet_count  favorite_count
0              0  892420643555336193           8842           39492
1              1  892177421306343426           6480           33786
2              2  891815181378084864           4301           25445
3              3  891689557279858688           8925           42863
4              4  891327558926688256           9721           41016
5              5  891087950875897856           3240           20548
6              6  890971913173991426           2142           12053
7              7  890729181411237888          19548           66596
8              8  890609185150312448           4403           28187
9              9  890240255349198849           7684           32467
10            10  890006608113172480           7584           31127
11            11  889880896479866881           5116           28208
12            12  889665388333682689           8502           38745
13            13  889638837579907072           4705           27633
14            14  889531135344209921           2309           15329
15            15  889278841981685760           5635           25712
16            16  888917238123831296           4681           29555
17            17  888804989199671297           4535           26021
18            18  888554962724278272           3722           20267
19            19  888078434458587136           3637           22144
20            20  887705289381826560           5584           30690
21            21  887517139158093824          12053           46940
22            22  887473957103951883          18813           70007
23            23  887343217045368832          10713           34223
24            24  887101392804085760           6147           31045
25            25  886983233522544640           8045           35786
26            26  886736880519319552           3420           12286
27            27  886680336477933568           4597           22802
28            28  886366144734445568           3297           21488
29            29  886267009285017600              4             117
...          ...                 ...            ...             ...
2322        2322  666411507551481857            337             457
2323        2323  666407126856765440             43             113
2324        2324  666396247373291520             91             171
2325        2325  666373753744588802             99             194
2326        2326  666362758909284353            590             801
2327        2327  666353288456101888             76             228
2328        2328  666345417576210432            146             308
2329        2329  666337882303524864             96             203
2330        2330  666293911632134144            365             519
2331        2331  666287406224695296             71             152
2332        2332  666273097616637952             81             183
2333        2333  666268910803644416             37             108
2334        2334  666104133288665088           6835           14703
2335        2335  666102155909144576             15              81
2336        2336  666099513787052032             73             160
2337        2337  666094000022159362             78             168
2338        2338  666082916733198337             47             121
2339        2339  666073100786774016            173             334
2340        2340  666071193221509120             67             154
2341        2341  666063827256086533            230             494
2342        2342  666058600524156928             61             117
2343        2343  666057090499244032            146             304
2344        2344  666055525042405380            261             449
2345        2345  666051853826850816            877            1250
2346        2346  666050758794694657             60             136
2347        2347  666049248165822465             41             111
2348        2348  666044226329800704            147           

## （二）编程评估：深入探索

### 1. **`twitters` 表格**

In [11]:
twitters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [23]:
twitters.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [15]:
twitters.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [32]:
twitters[twitters.rating_numerator == 1776].text

979    This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh
Name: text, dtype: object

In [39]:
twitters.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [33]:
twitters[twitters.rating_denominator == 170].text

1120    Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv
Name: text, dtype: object

In [40]:
twitters[twitters.rating_denominator == 11].text

784     RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…
1068    After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ
1662     This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5
Name: text, dtype: object

In [38]:
twitters.iloc[784,:]

tweet_id                                                                                                                                                                                                           775096608509886464
in_reply_to_status_id                                                                                                                                                                                                             NaN
in_reply_to_user_id                                                                                                                                                                                                               NaN
timestamp                                                                                                                                                                                                   2016-09-11 22:20:06 +0000
source                                                                          

1、打分很“任性”啊，有些打分很奇怪，但看内容确实如此  
2、提取text时候出现了个别错误，如rating_denominator基本应为10，但提取时候出现了明显错误，估计re表达式有问题

### 2. **`images` 表格**

In [12]:
images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [42]:
images.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [43]:
images[images.img_num == 4]

tweet_id                                          jpg_url  \
144   668623201287675904  https://pbs.twimg.com/media/CUdtP1xUYAIeBnE.jpg   
779   689905486972461056  https://pbs.twimg.com/media/CZMJYCRVAAE35Wk.jpg   
1024  710588934686908417  https://pbs.twimg.com/media/CdyE2x1W8AAe0TG.jpg   
1161  734787690684657664  https://pbs.twimg.com/media/CjJ9gQ1WgAAXQtJ.jpg   
1286  750868782890057730  https://pbs.twimg.com/media/CmufLLsXYAAsU0r.jpg   
1325  756998049151549440  https://pbs.twimg.com/media/CoFlsGAWgAA2YeV.jpg   
1337  758405701903519748  https://pbs.twimg.com/media/CoZl9fXWgAMox0n.jpg   
1342  758854675097526272  https://pbs.twimg.com/media/Cof-SuqVYAAs4kZ.jpg   
1372  762464539388485633  https://pbs.twimg.com/media/CpTRc4DUEAAYTq6.jpg   
1437  773985732834758656  https://pbs.twimg.com/media/Cr2_6R8WAAAUMtc.jpg   
1496  783391753726550016  https://pbs.twimg.com/media/Ct8qn8EWIAAk9zP.jpg   
1542  791312159183634433  https://pbs.twimg.com/media/CvtONV4WAAAQ3Rn.jpg   
1543  791406955684368384  https://pbs.twimg.com/media/CvukbEkWAAAV-69.jpg   
1549  792913359805018113  https://pbs.twimg.com/media/CwD-eCLWIAA6v0B.jpg   
1617  802323869084381190  https://pbs.twimg.com/media/CyJtSmDUAAA2F9x.jpg   
1713  818614493328580609  https://pbs.twimg.com/media/C1xNgraVIAA3EVb.jpg   
1716  819006400881917954  https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1717  819015331746349057  https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1766  826848821049180160  https://pbs.twimg.com/media/C3mOnZ_XUAAjr2V.jpg   
1768  827199976799354881  https://pbs.twimg.com/media/C3rN-lcWEAA9CmR.jpg   
1790  830097400375152640  https://pbs.twimg.com/media/C4UZLZLWYAA0dcs.jpg   
1795  831315979191906304  https://pbs.twimg.com/media/C4lst0bXAAE6MP8.jpg   
1800  831911600680497154  https://pbs.twimg.com/media/C4uLLGuUoAAkIHm.jpg   
1829  836001077879255040  https://pbs.twimg.com/media/C5oSiskU0AE8sJ_.jpg   
1893  849412302885593088  https://pbs.twimg.com/media/C8m3-iQVoAAETnF.jpg   
1914  854120357044912130  https://pbs.twimg.com/media/C9px7jyVwAAnmwN.jpg   
1920  856282028240666624  https://pbs.twimg.com/media/C-If9ZwXoAAfDX2.jpg   
1934  859851578198683649  https://pbs.twimg.com/media/C-7OcfyXsAAsqzU.jpg   
1954  864197398364647424  https://pbs.twimg.com/media/C_4-8iPV0AA1Twg.jpg   
1978  870656317836468226  https://pbs.twimg.com/media/DBUxSSTXsAA-Jn1.jpg   
2040  885167619883638784  https://pbs.twimg.com/media/DEi_N9qXYAAgEEw.jpg   

      img_num                        p1   p1_conf  p1_dog  \
144         4                 Chihuahua  0.708163    True   
779         4                Pomeranian  0.943331    True   
1024        4                  Pembroke  0.982004    True   
1161        4          golden_retriever  0.883991    True   
1286        4                toy_poodle  0.912648    True   
1325        4          golden_retriever  0.678555    True   
1337        4  Chesapeake_Bay_retriever  0.702954    True   
1342        4                    barrow  0.974047   False   
1372        4                      chow  0.999953    True   
1437        4               giant_panda  0.451149   False   
1496        4        Norwegian_elkhound  0.877130    True   
1542        4        miniature_pinscher  0.892925    True   
1543        4                  Pembroke  0.972629    True   
1549        4                  web_site  0.226716   False   
1617        4              home_theater  0.765069   False   
1713        4                 Chihuahua  0.450722    True   
1716        4                    prison  0.907083   False   
1717        4                    prison  0.907083   False   
1766        4            Great_Pyrenees  0.858764    True   
1768        4                Great_Dane  0.869681    True   
1790        4                toy_poodle  0.442713    True   
1795        4                    briard  0.982755    True   
1800        4                bloodhound  0.777562    True   
1829        4                   Samoyed  0.963558    True   
1893        4             

### 3. **`twitter_additions` 表格**

In [13]:
twitter_additions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2352 entries, 0 to 2351
Data columns (total 4 columns):
Unnamed: 0        2352 non-null int64
tweet_ID          2352 non-null int64
retweet_count     2352 non-null int64
favorite_count    2352 non-null int64
dtypes: int64(4)
memory usage: 73.6 KB


In [44]:
twitter_additions.describe()

Unnamed: 0      tweet_ID  retweet_count  favorite_count
count  2352.000000  2.352000e+03    2352.000000     2352.000000
mean   1175.500000  7.425913e+17    3134.932398     8109.198980
std     679.108239  6.846210e+16    5237.846296    11980.795669
min       0.000000  6.660209e+17       0.000000        0.000000
25%     587.750000  6.783949e+17     618.000000     1417.000000
50%    1175.500000  7.193536e+17    1456.500000     3596.500000
75%    1763.250000  7.991219e+17    3628.750000    10118.000000
max    2351.000000  8.924206e+17   79116.000000   132318.000000

In [46]:
twitter_additions[twitter_additions.retweet_count > twitter_additions.favorite_count]

Unnamed: 0            tweet_ID  retweet_count  favorite_count
31            31  886054160059072513            106               0
35            35  885311592912609280          19188               0
67            67  879130579576475649           7118               0
72            72  878404777348136964           1338               0
73            73  878316110768087041           6925               0
77            77  877611172832227328             82               0
90            90  874434818259525634          15442               0
94            94  873697596434513921          12435               0
96            96  873337748698140672           1656               0
100          100  872668790621863937             31               0
108          108  871166179821445120           5956               0
122          122  868639477480148993           2219               0
128          128  867072653475098625            134               0
130          130  866816280283807744          32705               0
135          135  866094527597207552           8882               0
144          144  863471782782697472           2664               0
156          156  860981674716409858           2317               0
157          157  860924035999428608            879               0
162          162  860177593139703809          33231               0
168          168  858860390427611136           8762               0
177          177  857062103051644929            182               0
179          179  856602993587888130          11580               0
182          182  856330835276025856            729               0
191          191  855245323840757760           6521               0
192          192  855138241867124737             50               0
201          201  852936405516943360           2226               0
208          208  851953902622658560          10647               0
209          209  851861385021730816             23               0
219          219  849668094696017920           5490               0
227          227  847978865427394560           3634               0
...          ...                 ...            ...             ...
775          775  775898661951791106          17522               0
790          790  773336787167145985           5860               0
796          796  772615324260794368           3881               0
807          807  771171053431250945           8653               0
811          811  771004394259247104            252               0
814          814  770743923962707968          52101               0
818          818  770093767776997377           3492               0
822          822  769335591808995329           8780               0
825          825  768909767477751808           3103               0
829          829  768554158521745409           6673               0
837          837  766864461642756096           6481               0
843          843  766078092750233600           2955               0
856          856  763167063695355904           3471               0
864          864  761750502866649088           4511               0
868          868  761371037149827077          20400               0
881          881  760153949710192640             38               0
886          886  759566828574212096          24183               0
891          891  759159934323924993           1348               0
904          904  757729163776290825           9247               0
907          907  757597904299253760            331               0
922          922  754874841593970688           9140               0
933          933  753298634498793472           6576               0
939          939  752701944171524096           3273               0
945          945  752309394570878976          18880               0
1008        1008  747242308580548608           3244               0
1019        1019  746521445350707200           1103               0
1039        1039  743835915802583040           2372           

怀疑没有点赞数、却有超高转发数的twitter是雇了“水军”的，哈哈~

## （三）评估结果：问题记录
### 1. 数据质量
**`twitters` 表格**
- id类的列应是字符串，不是数字（tweet_id、in_reply_to_status_id、in_reply_to_user_id、retweeted_status_id、retweeted_status_user_id）
- timestamp 中的后6位即' +0000'没有意义
- source应为iphone等，其它标签意义不大
- text正文应和短网址分开，拆分为正文和网址两列
- retweeted_status_id、retweeted_status_user_id、retweeted_status_timestamp数据大范围缺失**（无法清理）**
- expanded_urls小部分数据缺失**（无法清理）**
- rating_numerator个别值远超过10，如960，有的自动提取文字时出现错误，如911；文字提取错误的情况也出现在rating_denominator、name中
- rating_numerator出现无效分（0分）
- name、doggo、floofer、pupper、puppo中存在空值（NaN），原因是个别行从retweeted_status_id开始出现“错列”填写情况，向左1-3列或向右1列，至少98个

**`images` 表格**
- P1、P2、P3的字符中字母大小写不统一
- 最大概率P1为False，不属于狗
- p1_conf、p2_conf、p3_conf位数保留不一致，保留到小数点后2位即可
- 并没有p4以及相关的列，即img_num列中的4没有意义，检验后若按最大概率，全部可取1

**`twitter_additions` 表格**
- tweet_ID应是字符串，不是数字
- favorite_count部分为0**（无法清理）**

### 2. 数据结构
- twitter_additions、images应和twitters表格合并
- `twitters` 表的text列中的正文内容应和短网址分开，拆分为正文和网址两列
- doggo、floofer、pupper、puppo含义一致，应合并为一列stage

# 三、清理数据

以上共包含18个问题，对其中12个质量问题、3个结构问题进行处理。  
按数据整理维度，对上述问题进行归类并做适当合并。


Tips：
- 首先copy数据
- **define-code-test**框架应根据问题反复使用
- 对相似问题可以合并处理
- 按“缺失数据、结构问题、质量问题”的顺序依次处理问题
- 清洗完成后，导出形成主表`twitter_archive_master.csv`

In [75]:
twitters_copy = twitters.copy()
images_copy = images.copy()
twitter_additions_copy = twitter_additions.copy()

## （一）结构问题：补充表images、twitter_additions到表twitters中  
### 1. 定义
- 用pandas的[表合并](https://pandas.pydata.org/pandas-docs/stable/merging.html)操作，按tweet_id将表`images`、`twitter_additions`与当前主表`twitters`合并，形成新主表`df_master`
- 将id类列转换为str格式
- 导出主表，保存为`twitter_archive_master.csv`  

### 2. 代码  

In [76]:
twitters_copy.set_index('tweet_id', inplace=True)
images_copy.set_index('tweet_id',  inplace=True)
twitter_additions_copy.set_index('tweet_ID', inplace=True)

In [77]:
df_master = pd.concat([twitters_copy, images_copy, twitter_additions_copy], axis=1, join='inner')
df_master.drop(columns=['Unnamed: 0'], inplace=True)

In [45]:
df_master.index = df_master.index.astype('str')

In [102]:
id_col_names = ['in_reply_to_user_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id','in_reply_to_status_id']
df_master.loc[:, id_col_names] = df_master.loc[:, id_col_names].astype('str')

In [42]:
# 将表中所有的None值替换为NaN
df_master.replace('None', np.nan, inplace=True)

### 3. 测试

In [47]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2073 entries, 892420643555336193 to 666020888022790149
Data columns (total 30 columns):
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           2073 non-null object
timestamp                     2073 non-null object
source                        2073 non-null object
text                          2073 non-null object
retweeted_status_id           2073 non-null object
retweeted_status_user_id      2073 non-null object
retweeted_status_timestamp    79 non-null object
expanded_urls                 2073 non-null object
rating_numerator              2073 non-null int64
rating_denominator            2073 non-null int64
name                          1496 non-null object
doggo                         80 non-null object
floofer                       8 non-null object
pupper                        222 non-null object
puppo                         24 non-null object
jpg_url                       2073 non-null object
img_num      

In [31]:
# 检查主表是否有重复
df_master[df_master.duplicated()]

Empty DataFrame
Columns: [in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, retweet_count, favorite_count, twitter_urls]
Index: []

[0 rows x 30 columns]

## （二）结构问题：text列中的正文内容应和短网址分开，拆分为正文和网址两列  
### 1. 定义
- 用正则表达式和pandas的`str.extract`方法，从表df_master的列*text*中提取提取网址，保存为新列twitter_urls
- 用pandas的`str.split`方法，从列text中去掉网址

### 2. 代码  

In [2]:
import re

In [24]:
df_master['twitter_urls'] = df_master.text.str.extract(r'(https:.*[A-Z])' , expand=True)

In [25]:
df_master['text'] = df_master.text.str.split('https://', expand=True)[0]

### 3. 测试

In [26]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 892420643555336193 to 666020888022790149
Data columns (total 30 columns):
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2073 non-null object
source                        2073 non-null object
text                          2073 non-null object
retweeted_status_id           79 non-null float64
retweeted_status_user_id      79 non-null float64
retweeted_status_timestamp    79 non-null object
expanded_urls                 2073 non-null object
rating_numerator              2073 non-null int64
rating_denominator            2073 non-null int64
name                          2073 non-null object
doggo                         2073 non-null object
floofer                       2073 non-null object
pupper                        2073 non-null object
puppo                         2073 non-null object
jpg_url                       2073 non-null object
img

In [52]:
df_master

in_reply_to_status_id in_reply_to_user_id  \
892420643555336193                    NaN                 nan   
892177421306343426                    NaN                 nan   
891815181378084864                    NaN                 nan   
891689557279858688                    NaN                 nan   
891327558926688256                    NaN                 nan   
891087950875897856                    NaN                 nan   
890971913173991426                    NaN                 nan   
890729181411237888                    NaN                 nan   
890609185150312448                    NaN                 nan   
890240255349198849                    NaN                 nan   
890006608113172480                    NaN                 nan   
889880896479866881                    NaN                 nan   
889665388333682689                    NaN                 nan   
889638837579907072                    NaN                 nan   
889531135344209921                    NaN                 nan   
889278841981685760                    NaN                 nan   
888917238123831296                    NaN                 nan   
888804989199671297                    NaN                 nan   
888554962724278272                    NaN                 nan   
888078434458587136                    NaN                 nan   
887705289381826560                    NaN                 nan   
887517139158093824                    NaN                 nan   
887473957103951883                    NaN                 nan   
887343217045368832                    NaN                 nan   
887101392804085760                    NaN                 nan   
886983233522544640                    NaN                 nan   
886736880519319552                    NaN                 nan   
886680336477933568                    NaN                 nan   
886366144734445568                    NaN                 nan   
886258384151887873                    NaN                 nan   
...                                   ...                 ...   
666411507551481857                    NaN                 nan   
666407126856765440                    NaN                 nan   
666396247373291520                    NaN                 nan   
666373753744588802                    NaN                 nan   
666362758909284353                    NaN                 nan   
666353288456101888                    NaN                 nan   
666345417576210432                    NaN                 nan   
666337882303524864                    NaN                 nan   
666293911632134144                    NaN                 nan   
666287406224695296                    NaN                 nan   
666273097616637952                    NaN                 nan   
666268910803644416                    NaN                 nan   
666104133288665088                    NaN                 nan   
666102155909144576                    NaN                 nan   
666099513787052032                    NaN                 nan   
666094000022159362                    NaN                 nan   
666082916733198337                    NaN                 nan   
666073100786774016                    NaN                 nan   
666071193221509120                    NaN                 nan   
666063827256086533                    NaN                 nan   
666058600524156928                    NaN                 nan   
666057090499244032                    NaN                 nan   
666055525042405380                    NaN                 nan   
666051853826850816                    NaN                 nan   
666050758794694657                    NaN                 nan   
666049248165822465                    NaN                 nan   
666044226329800704                    NaN                 nan   
666033412701032449                    NaN                 nan   
666029285002620928                    NaN                 nan   
666020888022790149                    NaN              

## （三）结构问题：doggo、floofer、pupper、puppo应合并为一列stage
### 1.定义
- 用`str.cat`方法从主表中提取列*doggo、floofer、pupper、puppo*的值，合并形成新的列*stage*  
- 删除列*doggo、floofer、pupper、puppo*  

### 2.代码

In [92]:
df_master['stage'] = df_master['doggo'].str.cat([df_master['floofer'], df_master['pupper'], df_master['puppo']], na_rep='')

In [95]:
df_master.drop(columns=['doggo', 'floofer', 'pupper', 'puppo'], inplace=True)

### 3.测试

In [96]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2073 entries, 892420643555336193 to 666020888022790149
Data columns (total 27 columns):
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           2073 non-null object
timestamp                     2073 non-null object
source                        2073 non-null object
text                          2073 non-null object
retweeted_status_id           2073 non-null object
retweeted_status_user_id      2073 non-null object
retweeted_status_timestamp    79 non-null object
expanded_urls                 2073 non-null object
rating_numerator              2073 non-null int64
rating_denominator            2073 non-null int64
name                          1496 non-null object
jpg_url                       2073 non-null object
img_num                       2073 non-null int64
p1                            2073 non-null object
p1_conf                       2073 non-null float64
p1_dog                        2073 non-null bool
p2     

## （四）有效性问题：缺少p4以及相关的列
### 1.定义
提供的图像预测数据中没有p4等列，即img_num列中的4没有意义，检验后若按最大概率，可将4全部用1替换处理
### 2.代码

In [106]:
df_master.img_num.replace(4,1,inplace=True)

### 3.测试

In [100]:
df_master.img_num.value_counts()

1    1811
2     196
3      66
Name: img_num, dtype: int64

## （五）有效性问题：列*rating_numerator、rating_denominator*出现过大或过小值
### 1.定义
- 用公式：`rating_10 = rating_numerator / rating_denominator *10`重新计算，形成新列*rating_10*，该列的基准分全部为10
- 删除一些不合理的数据
- 删除列*rating_numerator、rating_denominator*  

### 2.代码

In [114]:
# 再次确认下打分是如何形成的
df_master[df_master.rating_denominator > 10].loc[:, ['text','rating_denominator']]

text  \
820690176645140481                                                     The floofs have been released I repeat the floofs have been released. 84/70    
758467244762497024                                                                          Why does this never happen at my front door... 165/150    
740373189193256964            After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP    
731156023742988288                               Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once    
722974582966214656                                                                                        Happy 4/20 from the squad! 13/10 for all    
716439118184652801                                   This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10    
713900603437621249                                                           Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody    
710658690886586372                                            Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80    
709198395643068416  From left to right:\r\nCletus, Jerome, Alejandro, Burp, &amp; Titson\r\nNone know where camera is. 45/50 would hug all at once    
704054845121142784                                                                      Here is a whole flock of puppers.  60/50 I'll take the lot    
697463031882764288                                                            Happy Wednesday here's a bucket of pups. 44/40 would pet all at once    
684225744407494656                          Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you    
684222868335505415            Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110    
682962037429899265             This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10    
677716515794329600                                                                                  IT'S PUPPERGEDDON. Total of 144/120 ...I think    
675853064436391936                                             Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once    

                    rating_denominator  
820690176645140481                  70  
758467244762497024                 150  
740373189193256964                  11  
731156023742988288                 170  
722974582966214656                  20  
716439118184652801                  50  
713900603437621249                  90  
710658690886586372                  80  
709198395643068416                  50  
704054845121142784                  50  
697463031882764288                  40  
684225744407494656                 130  
684222868335505415                 110  
682962037429899265                  11  
677716515794329600                 120  
675853064436391936                  80

In [113]:
len(df_master[df_master.rating_denominator > 10])

16

In [115]:
df_master['rating_10'] = df_master.rating_numerator / df_master.rating_denominator * 10

In [119]:
df_master['rating_10'].value_counts()

12.000000      478
10.000000      431
11.000000      419
13.000000      282
9.000000       151
8.000000        95
7.000000        51
14.000000       40
5.000000        35
6.000000        32
3.000000        19
4.000000        15
2.000000        10
1.000000         4
0.000000         2
420.000000       1
1776.000000      1
34.285714        1
26.000000        1
15.000000        1
27.000000        1
6.363636         1
8.181818         1
75.000000        1
Name: rating_10, dtype: int64

In [124]:
df_master[df_master['rating_10'] >= 20].text

810984652412424192                Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \r\nKeep Sam smiling by clicking and sharing this link:\r\n
786709082849828864                                     This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 
778027034220126208    This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back 
749981277374128128                                                                                       This is Atticus. He's quite simply America af. 1776/10 
680494726643068929                                                           Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 
670842764863651840                                                                              After so many requests... here you go.\r\n\r\nGood dogg. 420/10 
Name: text, dtype: object

In [128]:
df_master.drop(df_master[df_master['rating_10'] >= 20].index,inplace=True)

In [130]:
df_master.drop(['rating_numerator', 'rating_denominator'],axis=1, inplace=True)

### 3.测试

In [129]:
df_master['rating_10'].describe()

count    2067.000000
mean       10.592426
std         2.182066
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max        15.000000
Name: rating_10, dtype: float64

In [131]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2067 entries, 892420643555336193 to 666020888022790149
Data columns (total 26 columns):
in_reply_to_status_id         2067 non-null object
in_reply_to_user_id           2067 non-null object
timestamp                     2067 non-null object
source                        2067 non-null object
text                          2067 non-null object
retweeted_status_id           2067 non-null object
retweeted_status_user_id      2067 non-null object
retweeted_status_timestamp    79 non-null object
expanded_urls                 2067 non-null object
name                          1492 non-null object
jpg_url                       2067 non-null object
img_num                       2067 non-null int32
p1                            2067 non-null object
p1_conf                       2067 non-null float64
p1_dog                        2067 non-null bool
p2                            2067 non-null object
p2_conf                       2067 non-null float64
p2_

## （六）有效性问题：列*p1_dog*为False，不属于狗
### 1.定义
- 用pandas筛选出列*p1_dog*为False以及列*p1_conf*足够高的行，判定为不属于狗的照片

### 2.代码

In [149]:
df_master[(df_master.p1_dog == False) & (df_master.p1_conf > 0.98)].jpg_url

849051919805034497    https://pbs.twimg.com/media/C8hwNxbXYAAwyVG.jpg
844979544864018432    https://pbs.twimg.com/media/C7n4aQ0VAAAohkL.jpg
840370681858686976    https://pbs.twimg.com/media/C6mYrK0UwAANhep.jpg
838916489579200512    https://pbs.twimg.com/media/C6RkiQZUsAAM4R4.jpg
834167344700198914    https://pbs.twimg.com/media/C5OOxY6WAAAxERz.jpg
826115272272650244    https://pbs.twimg.com/media/C3bzVILWcAUjS5i.jpg
825876512159186944    https://pbs.twimg.com/media/C3YaSnQWAAILgz0.jpg
823322678127919110    https://pbs.twimg.com/media/C20HmaKWgAQ6-6X.jpg
820078625395449857    https://pbs.twimg.com/media/C2GBJADWIAQvcNb.jpg
818536468981415936    https://pbs.twimg.com/media/C1wGkYoVQAAuC_O.jpg
792883833364439040    https://pbs.twimg.com/media/CwDjoH3WAAIniIs.jpg
790581949425475584    https://pbs.twimg.com/media/Cvi2FiKWgAAif1u.jpg
781524693396357120    https://pbs.twimg.com/media/CtiIj0AWcAEBDvw.jpg
776088319444877312    https://pbs.twimg.com/media/CsU4NKkW8AUI5eG.jpg
775729183532220416  

![666362758909284353](https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg)

In [150]:
df_master.drop(df_master[(df_master.p1_dog == False) & (df_master.p1_conf > 0.98)].index, inplace=True)

### 3.测试

In [151]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2013 entries, 892420643555336193 to 666020888022790149
Data columns (total 26 columns):
in_reply_to_status_id         2013 non-null object
in_reply_to_user_id           2013 non-null object
timestamp                     2013 non-null object
source                        2013 non-null object
text                          2013 non-null object
retweeted_status_id           2013 non-null object
retweeted_status_user_id      2013 non-null object
retweeted_status_timestamp    76 non-null object
expanded_urls                 2013 non-null object
name                          1465 non-null object
jpg_url                       2013 non-null object
img_num                       2013 non-null int32
p1                            2013 non-null object
p1_conf                       2013 non-null float64
p1_dog                        2013 non-null bool
p2                            2013 non-null object
p2_conf                       2013 non-null float64
p2_

## （七）准确性问题
### 1.定义
- 列rating_10出现少量0值等极低值，用pandas选择后删除

### 2.代码

In [155]:
df_master[df_master.rating_10 <= 2].jpg_url

835152434251116546    https://pbs.twimg.com/media/C5cOtWVWMAEjO5p.jpg
746906459439529985    https://pbs.twimg.com/media/Cl2LdofXEAATl7x.jpg
722974582966214656    https://pbs.twimg.com/media/CgiFjIpWgAA4wVp.jpg
678675843183484930    https://pbs.twimg.com/media/CWskEqnWUAAQZW_.jpg
678424312106393600    https://pbs.twimg.com/media/CWo_T8gW4AAgJNo.jpg
675153376133427200    https://pbs.twimg.com/media/CV6gaUUWEAAnETq.jpg
670826280409919488    https://pbs.twimg.com/media/CU9A8ZuWsAAt_S1.jpg
670783437142401025    https://pbs.twimg.com/media/CU8Z-OxXAAA-sd2.jpg
668142349051129856    https://pbs.twimg.com/media/CUW37BzWsAAlJlN.jpg
667878741721415682    https://pbs.twimg.com/media/CUTILFiWcAE8Rle.jpg
666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg
666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg
666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg
Name: jpg_url, dtype: object

![666051853826850816](https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg)

In [156]:
df_master.drop(df_master[df_master.rating_10 <= 2].index, inplace=True)

### 3.测试

In [157]:
df_master.rating_10.describe()

count    2000.000000
mean       10.681273
std         2.017218
min         3.000000
25%        10.000000
50%        11.000000
75%        12.000000
max        14.000000
Name: rating_10, dtype: float64

## （八）一致性问题
### 1.定义
- 用str方法处理timestamp，去掉后6位即' +0000'
- 用pandas和正则表达式将列source提取出iphone等类型值
- 将P1、P2、P3字符中的字母统一为小写
- 将p1_conf、p2_conf、p3_conf位数全部保留到小数点后2位  

以上均采用pandas的str[相关方法](https://pandas.pydata.org/pandas-docs/stable/text.html#splitting-and-replacing-strings)处理

### 2.代码

In [4]:
# 处理列timestamp
df_master.timestamp = df_master.timestamp.str.split(' \+').str[0]

In [5]:
# 处理列source
df_master.source = df_master.source.str.extract(r'<.*>(.*)<.*>' , expand=True)

In [6]:
# 处理列P1、P2、P3、stage
list_col_lower = ['p1','p2','p3', 'stage']
for a in list_col_lower:
    df_master.loc[:, a] = df_master.loc[:, a].str.lower()

In [7]:
# 处理列p1_conf、p2_conf、p3_conf
list_col_round = ['p1_conf', 'p2_conf', 'p3_conf']
for b in list_col_round:
    df_master.loc[:, b] = df_master.loc[:, b].round(decimals=2)

### 3.测试

In [8]:
df_master.head()

in_reply_to_status_id  in_reply_to_user_id  \
892420643555336193                    NaN                  NaN   
892177421306343426                    NaN                  NaN   
891815181378084864                    NaN                  NaN   
891689557279858688                    NaN                  NaN   
891327558926688256                    NaN                  NaN   

                              timestamp              source  \
892420643555336193  2017-08-01 16:23:56  Twitter for iPhone   
892177421306343426  2017-08-01 00:17:27  Twitter for iPhone   
891815181378084864  2017-07-31 00:18:03  Twitter for iPhone   
891689557279858688  2017-07-30 15:58:51  Twitter for iPhone   
891327558926688256  2017-07-29 16:00:24  Twitter for iPhone   

                                                                                                                                                           text  \
892420643555336193                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10    
892177421306343426  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10    
891815181378084864                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10    
891689557279858688                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us    
891327558926688256  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek    

                    retweeted_status_id  retweeted_status_user_id  \
892420643555336193                  NaN                       NaN   
892177421306343426                  NaN                       NaN   
891815181378084864                  NaN                       NaN   
891689557279858688                  NaN                       NaN   
891327558926688256                  NaN                       NaN   

                   retweeted_status_timestamp  \
892420643555336193                        NaN   
892177421306343426                        NaN   
891815181378084864                        NaN   
891689557279858688                        NaN   
891327558926688256                        NaN   

                                                                                                                                      expanded_urls  \
892420643555336193                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
892177421306343426                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
891815181378084864                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
891689557279858688                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
891327558926688256  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

                        name    ...     p2_conf  p2_dog  \
892420643555336193   Phineas    ...        0.09   False   
892177421306343426     Tilly    ...        0.09    True   
891815181378084864    Archie    ...        0.08    True   
891689557279858688     Darla    ...        0.17    True   
891327558926688256  Franklin    ...        0.23    True   

                                             p3  p3_conf  p3_dog  \
892420643555336193                       banana     0.08   False   
892177421306343426                     papillon     0.07    True   
891815181378084864                       kelpie     0.03    True   
891689557279858688   

## 补充

In [3]:
df_master = pd.read_csv('twitter_archive_master.csv', index_col='Unnamed: 0')

In [9]:
df_master.to_csv('twitter_archive_master.csv')